### PR TITLE
extract contents fetching from jupyter-extension into core

### PR DESCRIPTION
**Pull request by rgbkrk**
Attempted to match much of the style from the kernelspecs setup. This doesn't change much other than set up similar wording of actions, constants, etc.

### CODE REVIEW

**Comment by rgbkrk** [comment]
Woohoo! Now the kernel is kicked off when the notebook is loaded. 😄

File: packages/core/src/actions.js [hunk 1]
```diff
@@ -84,11 +84,37 @@ import type {
   // TODO: Needs an action creator
   DoneSavingConfigAction,
   // TODO: Needs an action creator
-  SetNotebookCheckpointAction
+  SetNotebookCheckpointAction,
```
Comment (mpacer) [inline packages/core/src/actions.js:87]: Why did you add this here if you aren’t also adding that action?

File: packages/core/src/epics/kernel-lifecycle.js [hunk 2]
```diff
@@ -87,3 +92,36 @@ export const acquireKernelInfoEpic = (action$: ActionsObservable<*>) =>
     ofType(LAUNCH_KERNEL_SUCCESSFUL),
     switchMap(action => acquireKernelInfo(action.kernel.channels))
   );
+
+export const extractNewKernel = (
+  filename?: string,
+  notebook: ImmutableNotebook
+) => {
+  // TODO: There's some incongruence between desktop and web app here, regarding path vs. filename
+  const cwd = (filename && path.dirname(filename)) || "/";
```
Comment (mpacer) [inline packages/core/src/epics/kernel-lifecycle.js:101]: Why do we call it filename if it’s actually a path? Technically aren’t all filenames implicit (relative) paths while not all paths are filenames?  Wouldn’t filepath be the more appropriate name? This might make more sense in a later PR but is definitely worth considering.

Comment (mpacer) [reply packages/core/src/epics/kernel-lifecycle.js:101]: Also should this default to the root directory or is `/` here only being treated as a relative directory since it’s to be treated as a relative path and not an absolute path?

File: packages/core/src/epics/contents.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      const serverConfig = {
+        endpoint: host.serverUrl,
+        token: host.token,
+        crossDomain: false
+      };
+
+      return contents
+        .get(serverConfig, action.payload.path, action.payload.params)
+        .pipe(
+          tap(xhr => {
```
Comment (mpacer) [inline packages/core/src/epics/contents.js:49]: Not a problem, just I’m curious; does throwing an error inside `tap` propagate the error up and halt execution of the subsequent code inside the `pipe`? If so is the the common way to throw errors when using lettable operators?

File: packages/core/src/epics/kernel-lifecycle.js [hunk 4]
```diff
@@ -87,3 +92,36 @@ export const acquireKernelInfoEpic = (action$: ActionsObservable<*>) =>
...
+) =>
+  action$.pipe(
+    ofType(SET_NOTEBOOK),
+    map((action: SetNotebookAction) => {
+      const { cwd, kernelSpecName } = extractNewKernel(
+        action.filename,
+        action.notebook
+      );
+
+      return launchKernelByName(kernelSpecName, cwd);
```
Comment (mpacer) [inline packages/core/src/epics/kernel-lifecycle.js:125]: W00t!

File: packages/core/src/actions.js (hunk 1 above)
Comment (mpacer) [reply packages/core/src/actions.js:87]: Oh nvm, comma.

File: packages/core/src/actions.js [hunk 5]
```diff
@@ -84,11 +84,37 @@ import type {
...
+  FetchContent,
+  FetchContentFulfilled,
+  FetchContentFailed
```
Comment (mpacer) [inline packages/core/src/actions.js:90]: Is it possible to add a comma here for cleaner future diffs? Or is this a prettier opinion?

**Comment by mpacer** [comment]
Merging because the only thing that I don’t agree with is stylistic. Really nice & clear code!

File: packages/core/src/epics/contents.js (hunk 3 above)
Comment (rgbkrk) [reply packages/core/src/epics/contents.js:49]: No further operators will run in the pipe, only a `catchError` operator will catch the error thrown here.

File: packages/core/src/actions.js (hunk 5 above)
Comment (rgbkrk) [reply packages/core/src/actions.js:90]: That's a prettier thing.

File: packages/core/src/epics/kernel-lifecycle.js (hunk 2 above)
Comment (rgbkrk) [reply packages/core/src/epics/kernel-lifecycle.js:101]: I kept with the original nomenclature of this action as it was written, which was `filename`. We've got some incongruence between desktop and webapp here since the desktop app uses paths which many times need to be `path.resolve`d and the default should be their current working directory rather than `/`. In the case of the web app, `/` is the root of the notebook server.
